### PR TITLE
Multiclusterapp

### PR DESCRIPTION
**Pull request by sowmyav27**
@sangeethah please review

### CODE REVIEW

File: tests/v3_api/common.py [hunk 1]
```diff
@@ -24,6 +25,7 @@
 
 TEST_IMAGE = "sangeetha/mytestcontainer"
 CLUSTER_NAME = os.environ.get("RANCHER_CLUSTER_NAME", "")
+CLUSTER_NAME_2 = ""
```
Comment (sangeethah) [inline tests/v3_api/common.py]: @sowmyav27 CLUSTER_NAME_2 should be read from an environment variable similar to CLUSTER_NAME

File: tests/v3_api/common.py [hunk 2]
```diff
@@ -1102,3 +1104,66 @@ def validate_file_content(pod, content, filename):
     cmd_get_content = "/bin/bash -c 'cat {0}' ".format(filename)
     output = kubectl_pod_exec(pod, cmd_get_content)
     assert output.strip().decode('utf-8') == content
+
+def wait_for_mcapp_to_active(client, multiClusterApp, timeout=DEFAULT_MULTI_CLUSTER_APP_TIMEOUT):
+    time.sleep(5)
```
Comment (sangeethah) [inline tests/v3_api/common.py]: This sleep should not be needed

File: tests/v3_api/common.py [hunk 3]
```diff
@@ -1102,3 +1104,66 @@ def validate_file_content(pod, content, filename):
...
+    if CLUSTER_NAME == "" or CLUSTER_NAME_2 == "":
+        clusters = client.list_cluster().data
+    else:
+        clusters.append(client.list_cluster(name=CLUSTER_NAME).data)
+        clusters.append(client.list_cluster(name=CLUSTER_NAME_2).data)
+    assert len(clusters) == 2
+    return client, clusters
+
+
+def validate_multi_cluster_app_cluster(app_id1, app_id2, p_client1, p_client2):
```
Comment (sangeethah) [inline tests/v3_api/common.py]: This method can be moved to test_multi_cluster_app.py . We should be able to take in only McAppId as input to this method.

File: tests/v3_api/common.py [hunk 4]
```diff
@@ -1102,3 +1104,66 @@ def validate_file_content(pod, content, filename):
...
+        if time.time() - start > timeout:
+            raise AssertionError(
+                "Timed out waiting for state to get to active")
+        time.sleep(.5)
+        apps = client.list_app(name=app_id).data
+        assert len(apps) == 1
+        mapp = apps[0]
+    return mapp
+
+
```
Comment (sangeethah) [inline tests/v3_api/common.py]: This method can be moved to test_multi_cluster_app.py

File: tests/v3_api/common.py [hunk 5]
```diff
@@ -1102,3 +1104,66 @@ def validate_file_content(pod, content, filename):
...
+def get_admin_client_and_cluster_mcapp():
+    clusters = []
+    client = get_admin_client()
+    if CLUSTER_NAME == "" or CLUSTER_NAME_2 == "":
```
Comment (sangeethah) [inline tests/v3_api/common.py]: Change this to if CLUSTER_NAME and CLUSTER_NAME_2 are not empty , then use them else use any 2 clusters from the server.

File: tests/v3_api/common.py [hunk 6]
```diff
@@ -1102,3 +1104,66 @@ def validate_file_content(pod, content, filename):
...
+        clusters = client.list_cluster().data
+    else:
+        clusters.append(client.list_cluster(name=CLUSTER_NAME).data)
+        clusters.append(client.list_cluster(name=CLUSTER_NAME_2).data)
+    assert len(clusters) == 2
```
Comment (sangeethah) [inline tests/v3_api/common.py]: Change this to >= instead of ===

File: tests/v3_api/common.py (hunk 5 above)
Comment (sangeethah) [reply tests/v3_api/common.py]: Include error message for assert that MCApp tests can be run only on setups with atleast 2 clusters.

File: tests/v3_api/common.py (hunk 6 above)
Comment (sangeethah) [reply tests/v3_api/common.py]: Include error message for assert that MCApp tests can be run only on setups with atleast 2 clusters.

File: tests/v3_api/common.py [hunk 7]
```diff
@@ -1102,3 +1104,66 @@ def validate_file_content(pod, content, filename):
...
+
+
+def validate_mcapp_cluster(app_id, p_client):
+    mcapp = p_client.list_app(name=app_id).data
+    assert len(mcapp) == 1
+    app = mcapp[0]
+    return app
+
+
+def wait_for_mcapp_cluster_level_to_active(client, app_id, timeout=DEFAULT_MULTI_CLUSTER_APP_TIMEOUT):
```
Comment (sangeethah) [inline tests/v3_api/common.py]: Rename  this method to indicate that it for apps (not necessarily apps in MCAPP)

File: tests/v3_api/test_multi_cluster_app.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    p_client1 = namespace["namespace1"]["p_client"]
+    p_client2 = namespace["namespace2"]["p_client"]
+    multiClusterApp = client.create_multiClusterApp(templateVersionId=TEMP_VER,
+                                                    targets=targets,
+                                                    roles=ROLES,
+                                                    name=random_name(),
+                                                    answers=[answer])
+    multiClusterApp = wait_for_mcapp_to_active(client, multiClusterApp)
+    app_id1 = multiClusterApp.targets[0].appId
+    app_id2 = multiClusterApp.targets[1].appId
```
Comment (sangeethah) [inline tests/v3_api/test_multi_cluster_app.py]: should be able is directly assert for not None for app ids.

File: tests/v3_api/test_multi_cluster_app.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+    if app_id1 is None:
+        id1 = 1
+        print("app_id1 is None")
+    assert id1 == 0
+    if app_id2 is None:
+        id2 = 1
+        print("app_id2 is None")
+    assert id2 == 0
+    # verify if this app is available in the cluster/project
+    multiClusterApp = validate_multi_cluster_app_cluster(app_id1, app_id2,
```
Comment (sangeethah) [inline tests/v3_api/test_multi_cluster_app.py]: Include checks to make sure you are able to launch the wordress app as part of validation.

File: tests/v3_api/common.py (hunk 1 above)
Comment (sowmyav27) [reply tests/v3_api/common.py]: Added CLUSTER_NAME_2 = os.environ.get("RANCHER_CLUSTER_NAME_2", "")

File: tests/v3_api/common.py (hunk 5 above)
Comment (sowmyav27) [reply tests/v3_api/common.py]: Done.

File: tests/v3_api/common.py (hunk 6 above)
Comment (sowmyav27) [reply tests/v3_api/common.py]: Done.

File: tests/v3_api/common.py (hunk 3 above)
Comment (sowmyav27) [reply tests/v3_api/common.py]: Done

File: tests/v3_api/test_multi_cluster_app.py (hunk 8 above)
Comment (sowmyav27) [reply tests/v3_api/test_multi_cluster_app.py]: done

File: tests/v3_api/common.py (hunk 2 above)
Comment (sowmyav27) [reply tests/v3_api/common.py]: removed the sleep.

File: tests/v3_api/common.py (hunk 4 above)
Comment (sowmyav27) [reply tests/v3_api/common.py]: moved to the file.

File: tests/v3_api/common.py (hunk 7 above)
Comment (sowmyav27) [reply tests/v3_api/common.py]: done. Remaned to wait_for_mcapp_to_active()

**Comment by sangeethah** [comment]
@sowmyav27 Include version check as part of validate_multi_cluster_app_cluster()

**Comment by sangeethah** [comment]
LGTM

**Pull request merged**